packages/flutter_tools/lib/src/build_system/source.dart

```diff
@@ -166,7 +166,19 @@ class SourceVisitor implements ResolvedFiles {
   /// Visit a [Source] which is defined by an [Artifact] from the flutter cache.
   ///
   /// If the [Artifact] points to a directory then all child files are included.
+  /// To increase the performance of builds that use a known revision of Flutter,
+  /// these are updated to point towards the engine.version file instead of
+  /// the artifact itself.
   void visitArtifact(Artifact artifact, TargetPlatform platform, BuildMode mode) {
+    // This is not a local engine.
+    if (environment.engineVersion != null) {
+      sources.add(environment.flutterRootDir
+        .childDirectory('bin')
+        .childDirectory('internal')
+        .childFile('engine.version'),
```

Comment:
Technically, the contents of `//flutter/bin/cache` can vary based on other version files in `//flutter/bin/internal/*.version`, such as `gradle_wrapper.version`. Are any of these used during builds?

Reply:
No, only the engine.version so far. gradle.wrapper is just used to bootstrap the gradler worker, but that updates itself and delegates to the tool to do the actually dart work

Reply:
I don't understand. Are the .version files only updated in lockstep on an engine roll? Will this work on master if they're not updated in lockstep?

Reply:
Yes, the only version file used by assemble is engine.version and everything is updated in lockstep. On master this will work fine, since the hashes would have been the same anyways.

What this change does is makes a build after an engine upgrade slightly less efficient, because we assumed that all artifacts changed, in exchange for making every other build slightly more efficient.